WL4EditorWindow.cpp

```diff
@@ -1335,6 +1269,263 @@ void WL4EditorWindow::ClearEverythingInRoom(bool no_warning)
+/// </summary>
+/// <return>
+/// return false if the file cannot be loaded.
+/// </return>
+bool WL4EditorWindow::OpenRecentFile(QString filepath, const bool manageRecentScripts)
+{
+    // Check if it is a valid slot function call
+    if(filepath == "-/-" || filepath == "") return false;
+
+    int recentFileNum = recentROMnum;
+    int array_max_size = 5;
+    int array_start_id = 1;
+    if (manageRecentScripts)
+    {
+        recentFileNum = recentScriptNum;
+        array_start_id = 9;
+    }
+    QAction **actionlist_ptr = new QAction*[array_max_size];
+    if (!manageRecentScripts) {
+        for (int i = 0; i < array_max_size; i++) {
+            actionlist_ptr[i] = RecentROMs[i];
+        }
+    } else {
+        for (int i = 0; i < array_max_size; i++) {
+            actionlist_ptr[i] = RecentScripts[i];
+        }
+    }
+
+    // Check if the file exist, if not, modify the Recent ROM QAction list
+    QFile file(filepath);
+    if(!file.exists())
+    {
+        if(recentFileNum == 1)
+        {
+            SettingsUtils::SetKey(static_cast<SettingsUtils::IniKeys>(array_start_id), "");
+            actionlist_ptr[0]->setText("-/-");
+        }
+        if(recentFileNum > 1)
+        {
+            int deletelinenum = -1;
+            for(int i = 0; i < array_max_size; i++)
+            {
+                if(actionlist_ptr[i]->text() == filepath)
+                {
+                    deletelinenum = i;
+                    break;
+                }
+            }
+            for(int i = deletelinenum; i < (recentFileNum - 1); i++)
+            {
+                actionlist_ptr[i]->setText(actionlist_ptr[i + 1]->text());
+                SettingsUtils::SetKey(static_cast<SettingsUtils::IniKeys>(i + array_start_id), actionlist_ptr[i + 1]->text());
+            }
+            SettingsUtils::SetKey(static_cast<SettingsUtils::IniKeys>(recentFileNum + array_start_id - 1), "");
+            delete actionlist_ptr[recentFileNum - 1];
+        }
+        recentFileNum--;
+        QMessageBox::critical(nullptr, QString(tr("Load Error")), QString(tr("This File no longer exists!")));
+        return false;
+    }
+
+    // write back the value to the global variables
+    if (!manageRecentScripts) {
+        recentROMnum = recentFileNum;
+        for (int i = 0; i < array_max_size; i++)
+        {
+            RecentROMs[i] = actionlist_ptr[i];
+        }
+    } else {
+        recentScriptNum = recentFileNum;
+        for (int i = 0; i < array_max_size; i++)
+        {
+            RecentScripts[i] = actionlist_ptr[i];
+        }
+    }
+
+    return true;
+}
+
+/// <summary>
+/// a manager to deal with the recent ROM and recent script files' actions in the menu.
+/// </summary>
+void WL4EditorWindow::ManageRecentFilesOrScripts(QString newFilepath, const bool manageRecentScripts)
+{
+    int findedInRecentFile = -1; // start by 0
+    int recentFileNum = recentROMnum;
+    int array_max_size = 5;
```

Comment:
This should not be defined several times in functions, it should be in the H file as a const

Reply:
done